### ISSUE TITLE
Terminal UI buffer drops data?

### ISSUE THREAD

**Issue by iperry** [neovim/neovim#3030]
I'm attempting to run a GNU Make instance with termopen, and I'm seeing messages being dropped. I'm hooking into the on_exit() callback to write the terminal buffer to an external log, as well as parse it for errors for the quickfix list.

I deliberately added a compile error so that Make would error-out. When this happens, GCC prints to stderr, make exits, and my messages show up on the terminal. When I look at test.log, however, the last few lines are just _gone._

Rather than sharing the repository, I've managed to find a minimal case that results in lost data, namely by throwing lots of text at the terminal. I don't know if this is the same problem as the make issue, but with this test case, I get about 700 lines of "stdout," when I am expecting 2000.

I poked into it a bit and my guess is that data is being dropped somewhere after SIGCHLD is sent but there is still data either to be read in the TTY fds, or its already been sucked out of the fd but it makes it to the terminal buffer after the process has already exited and the exit callback has been called.

I've attached a minimal test case that fails on my computer as of 
5e9f9a875645af1e3c858daba799fe4a9021a767

Am I doing something stupid?

```
let s:Output = {}

function! s:Output.on_exit(id, code)
    write! test.log
endfunction

function! s:termtest()
    only
    vert botright sp | enew
    let s:Output['bufid'] = bufnr('%')
    call termopen([&sh, &shcf, './test.sh'], s:Output)
endfunction

command! -nargs=0 Termtest call s:termtest()
```

EDIT (@mhinz: shortened 2000 lines of code to two loops, as suggested by @lucc):

``` shell
#!/bin/sh
for i in {0..1000}; do echo "stdout $i"; done
for i in {0..1000}; do echo "stderr $i" 1>&2; done
```

**Comment by iperry** [comment]
Here's an interesting tidbit. I tried this exact same test on my Mac running Yosemite and it works fine. My failing system is Mint 17.2 at the exact same commit.

**Comment by lucc** [comment]
can you please change the shell script in your previous post to something shorter like
```bash
#!/bin/bash
for i in {0..1000}; do echo stdout $i; done
for i in {0..1000}; do echo stderr $i >&2; done
```
I assume that you should be able to observe the error with that script as well. Then you can remove the 2000 lines which just make this thread hard to read and provide no interesting info.

**Comment by mhinz** [comment]
1) Are you sure the buffer contains the correct output? Because the default scrollback size is 1000. Does the behaviour change with `:let g:terminal_scrollback_buffer_size = 2000`?

2) What happens if you add a `sleep 2` above the `write! test.log`?

**Comment by iperry** [comment]
@mhinz 
1) With test.sh, the buffer does not contain the correct output. There are lines missing. Changing the scrollback buffer size appears to have no affect. When I run my makefile job, the buffer actually does contain the correct output, but writing the file on_exit() has the last few lines are missing.

2) With test.sh, a sleep 2 has no effect, the buffer is still missing lines. With my makefile, a sleep 2 results in both a correct term buffer, and a log file with all of the output, including the previously missing lines.

**Comment by iperry** [comment]
I'm pretty sure it's dropping data somewhere between after the child exits and before we've read out all of the data from the pty. I can see 4095 bytes worth of output at a time being received. If the child exits quickly, I almost always lose the end of the terminal output. I can see EOF being sent in the libuv read_cb(), and looking at the byte counts it looks like EOF comes before all of the output has been pulled out of the fd. If I add a simple sleep 1 at the end of my script, everything turns out fine.

Running nvim in strace also seems to slow down the read_cb() calls enough that all the output is successfully received.

Maybe the shell itself is not flushing its output, or it closes its file descriptors without flushing output to the ptr, or maybe termios is doing something weird? I don't know enough about how the job system and libvte are put together to know how all the pipes and fds are plumbed together. Any ideas?

This all works perfectly on my Mac, for what it's worth.

**Comment by justinmk** [comment]
@iperry Could you see if https://github.com/neovim/neovim/pull/3029 fixes this issue?

**Comment by tarruda** [comment]
When a child exits(SIGCHLD is received), nvim will close the file descriptors connected to the child's stdio(master fd in case of pty jobs). This race happens when we close with data still waiting in the OS buffers, and AFAIK there's nothing we can do to prevent except by deferring the SIGCHLD handler as much as possible.

I'm aware of this problem and my one of my goals with #3029 is to enhance the event subsystem in a way that will be easier to predict and control the order various events are processed.

**Comment by justinmk** [comment]
Confirmed on latest master. Even if there's a race, it doesn't make sense that `test.log` is missing ~300 lines of output that _are_ present in the `:term` buffer. E.g. on my system `test.log` ends at `stderr 311` yet the `:term` buffer ends at `stderr 652`.

Users should be able to expect the `:term` buffer to `:write` reliably.

**Comment by oni-link** [comment]
Could be that the terminal buffer is not yet up to date when the exit-callback is called. 
Calling `refresth_terminal(...)` in `on_process_exit()` works for me. Most of the patch makes sure that no data is missing in the terminal. 
```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 47d44b1..35ade19 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -22157,6 +22157,9 @@ static void on_process_exit(Process *proc, int status, void *d)
 {
   TerminalJobData *data = d;
   if (data->term && !data->exited) {
+
+    refresh_terminal(data->term);
+
     data->exited = true;
     char msg[sizeof("\r\n[Process exited ]") + NUMBUFLEN];
     snprintf(msg, sizeof msg, "\r\n[Process exited %d]", proc->status);
diff --git a/src/nvim/event/process.c b/src/nvim/event/process.c
index 317e40e..aa04d78 100644
--- a/src/nvim/event/process.c
+++ b/src/nvim/event/process.c
@@ -352,31 +352,23 @@ static void flush_stream(Process *proc, Stream *stream)
   int err = uv_recv_buffer_size((uv_handle_t *)&stream->uv.pipe,
                                 &system_buffer_size);
   if (err) {
-    system_buffer_size = (int)rbuffer_capacity(stream->buffer);
+    // Max pipe size on Linux.
+    system_buffer_size = 1048576;
   }
 
   size_t max_bytes = stream->num_bytes + (size_t)system_buffer_size;
 
   // Read remaining data.
-  while (!stream->closed && stream->num_bytes < max_bytes) {
-    // Remember number of bytes before polling
-    size_t num_bytes = stream->num_bytes;
+  while (!stream->eof && stream->num_bytes < max_bytes && !stream->closed) {
 
     // Poll for data and process the generated events.
     loop_poll_events(proc->loop, 0);
+    if (rbuffer_size(stream->buffer) == 0) {
+        break;
+    }
     if (proc->events) {
         queue_process_events(proc->events);
     }
-
-    // Stream can be closed if it is empty.
-    if (num_bytes == stream->num_bytes) {
-      if (stream->read_cb) {
-        // Stream callback could miss EOF handling if a child keeps the stream
-        // open.
-        stream->read_cb(stream, stream->buffer, 0, stream->data, true);
-      }
-      break;
-    }
   }
 }
 
diff --git a/src/nvim/event/rstream.c b/src/nvim/event/rstream.c
index a520143..e202cf4 100644
--- a/src/nvim/event/rstream.c
+++ b/src/nvim/event/rstream.c
@@ -118,6 +118,7 @@ static void read_cb(uv_stream_t *uvstream, ssize_t cnt, const uv_buf_t *buf)
       // Read error or EOF, either way stop the stream and invoke the callback
       // with eof == true
       uv_read_stop(uvstream);
+      stream->eof = true;
       invoke_read_cb(stream, 0, true);
     }
     return;
@@ -181,6 +182,7 @@ static void read_event(void **argv)
     bool eof = (uintptr_t)argv[2];
     stream->read_cb(stream, stream->buffer, count, stream->data, eof);
   }
+
   stream->pending_reqs--;
   if (stream->closed && !stream->pending_reqs) {
     stream_close_handle(stream);
diff --git a/src/nvim/event/stream.c b/src/nvim/event/stream.c
index 3340415..49bfe12 100644
--- a/src/nvim/event/stream.c
+++ b/src/nvim/event/stream.c
@@ -72,6 +72,7 @@ void stream_init(Loop *loop, Stream *stream, int fd, uv_stream_t *uvstream,
   stream->buffer = NULL;
   stream->events = NULL;
   stream->num_bytes = 0;
+  stream->eof = false;
 }
 
 void stream_close(Stream *stream, stream_close_cb on_stream_close)
diff --git a/src/nvim/event/stream.h b/src/nvim/event/stream.h
index ad4e247..a9d0b05 100644
--- a/src/nvim/event/stream.h
+++ b/src/nvim/event/stream.h
@@ -50,6 +50,7 @@ struct stream {
   size_t maxmem;
   size_t pending_reqs;
   size_t num_bytes;
+  bool eof;
   void *data, *internal_data;
   bool closed;
   Queue *events;
diff --git a/src/nvim/rbuffer.c b/src/nvim/rbuffer.c
index a2cc432..14b652a 100644
--- a/src/nvim/rbuffer.c
+++ b/src/nvim/rbuffer.c
@@ -15,7 +15,7 @@ RBuffer *rbuffer_new(size_t capacity)
   FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_RET
 {
   if (!capacity) {
-    capacity = 0x10000;
+    capacity = 4095;
   }
 
   RBuffer *rv = xmalloc(sizeof(RBuffer) + capacity);
diff --git a/src/nvim/terminal.c b/src/nvim/terminal.c
index 6f50c03..53b0d31 100644
--- a/src/nvim/terminal.c
+++ b/src/nvim/terminal.c
@@ -952,7 +952,7 @@ static void invalidate_terminal(Terminal *term, int start_row, int end_row)
   }
 }
 
-static void refresh_terminal(Terminal *term)
+void refresh_terminal(Terminal *term)
 {
   // TODO(SplinterOfChaos): Find the condition that makes term->buf invalid.
   buf_T *buf = handle_get_buffer(term->buf_handle);

```

**Comment by radenling** [comment]
@justinmk I think the reason for `test.log` and the `:term` being out of sync is that `on_exit` is sometimes called with a stale terminal buffer. This should fix that: radenling/neovim@3a0c33e568316

**Comment by oni-link** [comment]
@radenling, do we want to have `[Process exited 0]` also written to the log file?

**Comment by radenling** [comment]
@oni-link, probably not. It's part of the buffer but it's not part of the program output. I'll fix that.

**Comment by radenling** [comment]
I think I've fixed this issue for linux (haven't tested it on any other platform and not sure the error occurs on any other platform or if I've broken something for other platforms). You can try [fix-libuv-stream](https://github.com/radenling/neovim/tree/fix-libuv-stream) if anyone wants to try it out.

**Comment by KillTheMule** [comment]
Awesome :+1: Could you create a PR? This way, it runs through all the CI tests, so you can see if it poses a problem on not-linux.

**Comment by justinmk** [comment]
For reference https://github.com/junegunn/fzf/issues/216#issuecomment-160353332 

> nvim does not update the terminal buffer as soon as it receives data. Here's a high-level overview of how it works:

> - Nvim receives data from the pty
> - Data is fed to libvterm, which immediately updates its internal screen
> - At most 10ms later, nvim updates the terminal buffer to reflect libvterm internal screen.

> In other words, nvim updates the buffer at most once each 10 milliseconds. This is done because updating buffers is more expensive than simply updating a virtual screen, and if we updated the buffer every time data is received, there could be serious performance problems with programs that send large bursts of output

**Comment by justinmk** [comment]
@oni-link @radenling Did `flush_stream()` in #4646 not address this issue?

**Comment by oni-link** [comment]
Without https://github.com/radenling/neovim/commit/9b9ad59dbd1e329a4de631a8ceeda3b7a83bb7d3#diff-148af05e3fe185d22c19b2ceb9ba5196
I still see missing output in an ASAN build when running the following script with `nvim -S issue.3030.vim +Termtest +Termtest`.

issue.3030.vim:
```vim
let s:Output = {}

function! s:termtest()
    only
    vert botright sp | enew
    let s:Output['bufid'] = bufnr('%')
    call termopen([&sh, &shcf, './issue.3030.sh'], s:Output)
endfunction

command! -nargs=0 Termtest call s:termtest()
```
issue.3030.sh:
```bash
#!/bin/bash

for i in {0..1000}; do echo "stdout $i"; done
for i in {0..1000}; do echo "stderr $i" 1>&2; done

```

**Comment by AckslD** [comment]
Not sure if this adds anything but if I have for example a python script:
```python
for i in range(10000):
    print(i)
```
and I run
```
:term python script.py
```
then the terminal buffer ends with something like:
```
...
8041
8042
8043
8044
804
[Process exited 0]
```

However if I open a terminal with `:term` and run `python script.py` in there I get all the lines.

**Comment by SamWindell** [comment]
I'm having this issue too, but I'm pretty out of my depth trying to work out how this works in the neovim source. It's quite a pesky issue because if for example you are running a compiler, it normally means the error text is being cut from the end of the output.

I can reproduce the exact case from the comment above this.

Here's some extra information for if someone takes a look.
- I'm using Linux Fedora, with neovim built from source at b649a96fc09fa7f18a988b0b0be5dadb54ad49de.
- I'm calling `vim.fn.termopen()`. The issue persists regardless of whether the I pass the command as a string or a list (docs say results in jobstart running in a shell or not).
- Issue is not there if I open a terminal buffer with `:terminal` and run the command there.
- Seems to work fine on Mac.
- If the command only outputs a small amount of text to stdout, this problem is not present. It seems to happen when there is a lot of text being passed to stdout and then the application abruptly ends. When it does happen, it happens every time.

**Comment by mikesmithgh** [comment]
I tested this with my Mac `15.1.1 (24B91)` and an Ubuntu github runner `ubuntu-24.04`.

No issues on Mac, nvim v0.10.03 (and nightly), python 3.11.4

![Image](https://github.com/user-attachments/assets/d3c87e98-558c-46b6-9443-e22b259d5c49)


Same problem as @AckslD for Ubuntu, nvim v0.10.03, python 3.12.3

<img width="767" alt="Image" src="https://github.com/user-attachments/assets/278468af-d5d2-432e-be66-66236aa9d45c" />

Similar errors are observed inconsistently and infrequently on test runs of kitty-scrollback.nvim with tmux integration https://github.com/mikesmithgh/kitty-scrollback.nvim/issues/115#issuecomment-2563004094.  

Adding a sleep does seem to workaround the issue:

<img width="767" alt="Image" src="https://github.com/user-attachments/assets/7ae499ab-4b5f-493c-875d-6e55d37a2527" />

**Comment by zeertzjq** [comment]
This can happen even for a PTY job without a terminal buffer:
```vim
let g:stdout = []
func OnOutput(id, data, event) dict
  let g:stdout += a:data
endfunc
call jobstart('seq 10000', {
      \ 'on_stdout': function('OnOutput'),
      \ 'pty': v:true,
      \ })
```
`g:stdout` sometimes ends before reaching `10000`.

This also seems unrelated to SIGCHLD. I did some testing and the cause seems to be that the `uv_read_cb` passed to `uv_read_start()` is too slow. The same problem happen with the following C program:
```c
#include <assert.h>
#include <fcntl.h>
#include <pty.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <sys/param.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <uv.h>

static void os_set_cloexec(const int fd)
{
  int fdflags = fcntl(fd, F_GETFD);
  if (fdflags < 0) {
    abort();
  }
  if ((fdflags & FD_CLOEXEC) == 0
      && fcntl(fd, F_SETFD, fdflags | FD_CLOEXEC) == -1) {
    abort();
  }
}

static void init_termios(struct termios *termios)
{
  termios->c_iflag = ICRNL|IXON;
  termios->c_oflag = OPOST|ONLCR;
#ifdef TAB0
  termios->c_oflag |= TAB0;
#endif
  termios->c_cflag = CS8|CREAD;
  termios->c_lflag = ISIG|ICANON|IEXTEN|ECHO|ECHOE|ECHOK;

  cfsetispeed(termios, 38400);
  cfsetospeed(termios, 38400);

#ifdef IUTF8
  termios->c_iflag |= IUTF8;
#endif
#ifdef NL0
  termios->c_oflag |= NL0;
#endif
#ifdef CR0
  termios->c_oflag |= CR0;
#endif
#ifdef BS0
  termios->c_oflag |= BS0;
#endif
#ifdef VT0
  termios->c_oflag |= VT0;
#endif
#ifdef FF0
  termios->c_oflag |= FF0;
#endif
#ifdef ECHOCTL
  termios->c_lflag |= ECHOCTL;
#endif
#ifdef ECHOKE
  termios->c_lflag |= ECHOKE;
#endif

  termios->c_cc[VINTR] = 0x1f & 'C';
  termios->c_cc[VQUIT] = 0x1f & '\\';
  termios->c_cc[VERASE] = 0x7f;
  termios->c_cc[VKILL] = 0x1f & 'U';
  termios->c_cc[VEOF] = 0x1f & 'D';
  termios->c_cc[VEOL] = _POSIX_VDISABLE;
  termios->c_cc[VEOL2] = _POSIX_VDISABLE;
  termios->c_cc[VSTART] = 0x1f & 'Q';
  termios->c_cc[VSTOP] = 0x1f & 'S';
  termios->c_cc[VSUSP] = 0x1f & 'Z';
#if !defined(__HAIKU__)
  termios->c_cc[VREPRINT] = 0x1f & 'R';
  termios->c_cc[VWERASE] = 0x1f & 'W';
  termios->c_cc[VLNEXT] = 0x1f & 'V';
#endif
  termios->c_cc[VMIN] = 1;
  termios->c_cc[VTIME] = 0;
}

static void set_duplicating_descriptor(int fd, uv_pipe_t *pipe)
{
  int fd_dup = dup(fd);
  if (fd_dup < 0) {
    abort();
  }

  os_set_cloexec(fd_dup);

  if (uv_pipe_open(pipe, fd_dup) != 0) {
    abort();
  }
}

static void alloc_cb(uv_handle_t *handle, size_t suggested_size, uv_buf_t *buf)
{
  buf->base = malloc(suggested_size);
  buf->len = suggested_size;
}

static void read_cb(uv_stream_t *uvstream, ssize_t cnt, const uv_buf_t *buf)
{
  if (cnt > 0) {
    write(STDOUT_FILENO, buf->base, (size_t)cnt);
    uv_sleep(1);  // Simulate slow computations.
  } else if (cnt < 0) {
    printf("\n%s\n", uv_strerror(cnt));
  }
  free(buf->base);
}

int main(void) {
  uv_loop_t loop;
  uv_loop_init(&loop);
  uv_pipe_t pipe_in;
  uv_pipe_init(&loop, &pipe_in, 0);
  uv_pipe_t pipe_out;
  uv_pipe_init(&loop, &pipe_out, 0);

  struct termios termios_default;
  init_termios(&termios_default);

  struct winsize winsize = { 24, 80, 0, 0 };
  uv_disable_stdio_inheritance();
  int master;
  int pid = forkpty(&master, NULL, &termios_default, &winsize);

  if (pid == 0) {
    execl("/usr/bin/seq", "/usr/bin/seq", "10000", NULL);
  }
  if (pid <= 0) {
    abort();
  }

  int master_status_flags = fcntl(master, F_GETFL);
  if (master_status_flags == -1) {
    abort();
  }
  if (fcntl(master, F_SETFL, master_status_flags | O_NONBLOCK) == -1) {
    abort();
  }

  os_set_cloexec(master);

  set_duplicating_descriptor(master, &pipe_in);
  set_duplicating_descriptor(master, &pipe_out);

  uv_read_start((uv_stream_t *)&pipe_out, alloc_cb, read_cb);
  uv_run(&loop, UV_RUN_DEFAULT);

  return EXIT_SUCCESS;
}
```
And if I remove the `uv_sleep(1)` in `read_cb()`, the data dropping doesn't happen.

I tried fixing this using the currently unused `events` field of `Stream`, which makes `invoke_read_cb()` defer `read_event` for a bit:
```diff
diff --git a/src/nvim/channel.c b/src/nvim/channel.c
index da002029e5..a9846e6ebf 100644
--- a/src/nvim/channel.c
+++ b/src/nvim/channel.c
@@ -451,6 +451,10 @@ Channel *channel_job_start(char **argv, const char *exepath, CallbackReader on_s
     rstream_start(&proc->err, on_job_stderr, chan);
   }
 
+  if (proc->type == kProcTypePty) {
+    proc->out.s.events = main_loop.fast_events;
+  }
+
   *status_out = (varnumber_T)chan->id;
   return chan;
 }
```
but it doesn't work either.

**Comment by zeertzjq** [comment]
Update: it appears that even without the `uv_sleep(1)`, the C program above can drop data about 10% of the times. Perhaps using `uv_pipe_t` to wrap PTY master is the problem (and `uv_poll_t` should be used instead)?

**Comment by zeertzjq** [comment]
This problem seems be result of what is described in <https://github.com/libuv/libuv/issues/2319>: libuv treats a short read after POLLHUP as EOF, but this isn't necessarily true for a PTY master device.

However, since a PTY master device is not a pipe, this specific case can't really be considered a libuv bug.

**Comment by zeertzjq** [comment]
Actually it's still possible that this is a libuv bug. I've reported this at <https://github.com/libuv/libuv/issues/4992> just in case.

If <https://github.com/libuv/libuv/issues/4992> is fixed for `uv_pipe_t` then this issue can be simply fixed by a small change to `flush_stream()`.
